consensus/tendermint/messages.go

```diff
@@ -0,0 +1,126 @@
+package tendermint
+
+// Todo: Signature over the messages needs to be handled somewhere. There are 2 options:
+//	1. Add the signature to each message and extend the Validator Set interface to include VerifyMessageSignature
+//	method.
+//	2. The P2P layer signs the message before gossiping to other validators and verifies the signature before passing
+//	the message to the consensus engine.
+//  The benefit of P2P layer handling the verification of the signature is the that the consensus layer can assume
+//  the message is from a validator in the validator set. However, this means that the P2P layer would need to be aware
+//  of the validator set and would need access to the blockchain which may not be a good idea.
+
+type Message[V Hashable[H], H Hash, A Addr] interface {
+	Proposal[V, H, A] | Prevote[H, A] | Precommit[H, A]
+}
+
+type Proposal[V Hashable[H], H Hash, A Addr] struct {
+	Height     uint
+	Round      uint
+	ValidRound *uint
```

Comment:
Nit: mybe this is too early for these types of comments, but it might be good to make a comment that nil means -1 or the round if not valid. Same with Value.

Reply:
I will remove them when the PR is ready, it was just present so that I don't forget why I made a certain decision and what the considerations were for that decision.

Not sure what you mean by nil means that no round reached a quorum. A nil validRound doesn't mean that.